src/s3path.jl

```diff
@@ -91,41 +93,40 @@ function S3Path(str::AbstractString; config::AbstractAWSConfig=global_aws_config
     result = tryparse(S3Path, str; config=config)
     result !== nothing || throw(ArgumentError("Invalid s3 path string: $str"))
     if version !== nothing && !isempty(version)
-        result.version !== nothing && throw(ArgumentError("Object `version` already parsed from `str`"))
+        if result.version !== nothing && result.version != version
+            throw(ArgumentError("Conflicting object versions in `version` and `str`"))
+        end
         result = S3Path(result.bucket, result.key; config=result.config, version=version)
     end
     return result
 end
 
 # if config=nothing, will not try to talk to AWS until after string is confirmed to be an s3 path
 function Base.tryparse(::Type{<:S3Path}, str::AbstractString; config::Union{Nothing,AbstractAWSConfig}=nothing)
-    str = String(str)
     startswith(str, "s3://") || return nothing
     # we do this here so that the `@p_str` macro only tries to call AWS if it actually has an S3 path
-    (config ≡ nothing) && (config = global_aws_config())
-    root = ""
-    path = ()
-    isdirectory = true
-
-    tokenized = split(str, "/")
-    bucket = strip(tokenized[3], '/')
-    drive = "s3://$bucket"
-
-    if length(tokenized) > 3
-        root = "/"
-        # If the last tokenized element is an empty string then we've parsed a directory
-        isdirectory = isempty(last(tokenized))
-        path = Tuple(filter!(!isempty, tokenized[4:end]))
-    end
-
-    return S3Path(path, root, drive, isdirectory, config, nothing)
+    config === nothing && (config = global_aws_config())
+    uri = URI(str)
```

Comment:
Minor improvement GitHub isn't letting me suggest
```julia
uri = URI(str)
uri.scheme == "s3" || return nothing

# we do this here so that the `@p_str` macro only tries to call AWS if it actually has an S3 path
config === nothing && (config = global_aws_config())
```